change xi macro to #% tag macro

gilch opened the pull request:
As discussed in #1439. This replaces `xi` with a more Clojure-like `#%` tag macro.

ekaschalk commented:
Do you have thoughts on whether to highlight the `%i`? I don't know how clojure does it.

gilch commented:
You mean, "does clojure-mode?" It does, I checked. Should Hy? I'm less certain. This is not builtin syntax. This is not even a core macro. 

But, if we conventionally used `%`-prefixed anaphors in other macros (as Clojure does), then highlighting such symbols would make sense. It does improve readability. But some might argue the old `xi` symbols were more readable, because that's what won out in #879, even though I presented `%` symbols as an option.

We could use `#xi` with the old `xi` symbols. Maybe it's more readable, but it's also more likely to interfere with user symbols. Names like `x1`, `x2` could totally come up.

ekaschalk commented:
I meant "does and what face", I can check into it.

A middle-ground is to implement is as an option, your call whether to enable it by default or not.

gilch commented:
I added a `%**` for the `&kwargs` parameter. This makes `#%` usable for partial application even if the function takes keyword arguments. I also changed the `&rest` parameter from `%&` to `%*` for consistency.

For example, `#%(foo #* %* #** %**)` would do the same thing as function `foo`, regardless of its signature.

Kodiologist commented:
This is cool. I like it. Is it ready for review? Still, I would finish up your outstanding PRs (and review those of others) before starting new ones like this.

gilch commented:
Hy can't be my top priority all the time. I'm busy with other things, but this seemed like a really easy fix, especially since I wrote the code in question in the first place.

It's a tag macro, so I've generalized `#%` to work on arbitrary expressions. 
For example, you can just use `#%[%1 %2]` instead of or
`(xi identity [x1 x2])` like before.

Clojure would have to use `#(identity [%1 %2])` or `#(vector %1 %2)` for that, but it saves the extra `%` in the more common case of a normal s-expression. We'd have to build function literals into the compiler to do it Clojure's way.

Kodiologist commented:
A lot of the diff is whitespace changes to old code. If you want to include them in this PR, could you put them in a separate commit?

Instead of "presence of a ``%`` parameter", I would write "presence of symbols named ``%*``, ``%**``, or ``%i`` for some number ``i``".

The tests seem to pass without the `(import hy)` in anaphoric.hy, despite the use of `HySymbol`. So I think you can remove it.

`(if (coll? expr) (flatten expr) (, expr))` can be written `(flatten [expr])`.

You have a commented-out `require` at the bottom of anaphoric.hy. I'm guessing it's old debugging code.

gilch commented:
> whitespace

Done.

> Instead of "presence

I reworded it, but not quite that way.

>  (import hy)

Yep.

> (flatten [expr])

More elegant, I like it.

> old debugging code

Removed.

@Kodiologist, how about now?

Kodiologist commented on docs/extra/anaphoric.rst:
```diff
@@ -229,21 +229,37 @@ Returns a function which applies several forms in series from left to right. The
+    => (add-10 6)
+    16
+
+``#%`` determines the parameter list by the presence of a ``%*``, or ``%**``
```
Stray comma.
